feat: cache balance history in DB

stefandunca opened the pull request:
Implement fetching of balance history from the database if it exists, otherwise, retrieves it using RPC calls. Every balance is stored in DB as an entry.

Also fetch regular balance history for all enabled networks, accounts and tokens. Fetching for a specific configuration will pause the full update which will resume after the specific configuration is fetched.

The granularity of data points returned is defined by the constant increment step defined in `timeIntervalToHoursPerStep` for each interval. The `timeIntervalToHoursPerStep` values have a common divisor to have a cache hit between requests.

To ensure cache hit, the requested blocks are "snapped" to specific time points defined by the `timeIntervalToSnapUnit` enumeration

The cached data is stored in `balance_history` table.
The "calibration" data, special case, used for bloc average timestamp is stored with address `0x0` and balance is not fetched/used.

Uniqueness constrained is enforced by the `balance_history_identify_entry` UNIQUE index.

Optimal DB fetching is ensured by the `balance_history_filter_entries` index

Tests are implemented using a mock of `BlockInfoSource` interface used to intercept the RPC calls.
`getBalanceHistoryFromBlocksSource` contains main functionality independent of the `chainClientSource`, `BlockInfoSource`'s concrete implementation.

Notes:

- the timestamp for a block balance is approximate and based on a very rough heuristic: blocks in between data points have a constant timestamp (average of the first and last block in the one-year rage)
- this is a complete refactoring of the logic used to fetch due to the nature of the heuristic chosen to ensure cache hit independent of the time interval if the granularity matches the cache
- moved balance history implementation from the `transaction.Controller` to `transaction.BalanceHistory` struct

Updates status-desktop #8226

stefandunca committed:
feat: add balance_history DB table and update migration scripts

Updates status-desktop #8226

stefandunca committed:
feat: cache to database retrieved balance history

Implement fetching of balance history from the database if it exists,
otherwise retrieves it using a RPC calls. Every balance is stored in DB
as an entry.

Also fetch regularly balance history for all enabled networks, accounts
and tokens. Fetching for a specific configuration will pause the
full update which will resume after the specific configuration is fetched.

The granularity of data points returned is defined by the constant
increment step define in `timeIntervalToHoursPerStep` for each interval.
The `timeIntervalToHoursPerStep` values have a common divisor to
have cache hit between requests.

To ensure cache hit the requested blocks are "snapped" to specific
time points defined by the `timeIntervalToSnapUnit` enumeration

The cached data is stored in `balance_history` table.
The "calibration" data, special case, used for bloc average timestamp is
stored with address `0x0` and balance is not fetched/used.

Uniqueness constrained is enforced by the `balance_history_identify_entry`
UNIQUE index.

Optimal DB fetching is ensured by the `balance_history_filter_entries`
index

Tests are implemented using a mock of `BlockInfoSource` interface used
to intercept the RPC calls.
`getBalanceHistoryFromBlocksSource` contains main functionality independent
of the `chainClientSource`, `BlockInfoSource`'s concrete implementation.

Notes:

- the timestamp for a block balance is approximate and based on a very
rough heuristic: blocks in between data points have a constant timestamp
(average of the fist and last block in the one year rage)
- this is a complete refactoring of the logic used to fetch due to the
nature of the heuristic chosen to ensure cache hit independent of the
time interval if the granularity matches the cache

Updates status-desktop #8226

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [x] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (85)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 3f611c0b | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/1/) | 2022-11-16 12:26:52 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221116-122215-3f611c-pr2962.zip) |
| :heavy_check_mark: | 3f611c0b | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/1/) | 2022-11-16 12:28:31 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221116-122215-3f611c-pr2962.zip) |
| :heavy_check_mark: | 3f611c0b | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/1/) | 2022-11-16 12:29:12 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221116-122215-3f611c-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | c835178b | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/2/) | 2022-11-16 12:32:32 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221116-122834-c83517-pr2962.zip) |
| :heavy_check_mark: | c835178b | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/2/) | 2022-11-16 12:32:38 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221116-122655-c83517-pr2962.zip) |
| :heavy_check_mark: | c835178b | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/2/) | 2022-11-16 12:35:23 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221116-122917-c83517-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | c82fe138 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/3/) | 2022-11-16 12:52:30 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221116-124911-c82fe1-pr2962.zip) |
| :heavy_check_mark: | c82fe138 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/3/) | 2022-11-16 12:54:18 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221116-124911-c82fe1-pr2962.zip) |
| :heavy_check_mark: | c82fe138 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/3/) | 2022-11-16 12:55:50 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221116-124911-c82fe1-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | cd4de462 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/4/) | 2022-11-16 17:27:42 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221116-172438-cd4de4-pr2962.zip) |
| :heavy_check_mark: | cd4de462 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/4/) | 2022-11-16 17:28:52 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221116-172438-cd4de4-pr2962.aar) |
| :heavy_check_mark: | cd4de462 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/4/) | 2022-11-16 17:29:28 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221116-172442-cd4de4-pr2962.zip) |
| | | | | | | |
| :heavy_check_mark: | 78b024b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/5/) | 2022-11-17 22:24:31 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221117-222221-78b024-pr2962.zip) |
| :heavy_check_mark: | 78b024b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/5/) | 2022-11-17 22:25:15 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221117-222221-78b024-pr2962.zip) |
| :heavy_check_mark: | 78b024b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/5/) | 2022-11-17 22:26:02 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221117-222221-78b024-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | b5d38929 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/6/) | 2022-11-18 23:14:10 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221118-231133-b5d389-pr2962.zip) |
| :heavy_check_mark: | b5d38929 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/6/) | 2022-11-18 23:15:39 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221118-231134-b5d389-pr2962.zip) |
| :heavy_check_mark: | b5d38929 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/6/) | 2022-11-18 23:16:57 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221118-231134-b5d389-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | 4d8b1414 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/7/) | 2022-12-05 19:15:57 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221205-191259-4d8b14-pr2962.zip) |
| :heavy_check_mark: | 4d8b1414 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/7/) | 2022-12-05 19:17:25 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221205-191259-4d8b14-pr2962.zip) |
| :heavy_check_mark: | 4d8b1414 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/7/) | 2022-12-05 19:18:40 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221205-191255-4d8b14-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | 718d215a | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/8/) | 2022-12-06 22:15:05 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221206-221231-718d21-pr2962.zip) |
| :heavy_check_mark: | 718d215a | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/8/) | 2022-12-06 22:15:36 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221206-221231-718d21-pr2962.zip) |
| :heavy_check_mark: | 718d215a | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/8/) | 2022-12-06 22:16:13 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221206-221228-718d21-pr2962.aar) |
| | | | | | | |
| :x: | c10132dd | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/9/) | 2022-12-07 19:38:06 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/9/consoleText) |
| :heavy_check_mark: | c10132dd | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/9/) | 2022-12-07 19:38:53 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221207-193655-c10132-pr2962.zip) |
| :heavy_check_mark: | c10132dd | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/9/) | 2022-12-07 19:39:04 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221207-193655-c10132-pr2962.zip) |
| | | | | | | |
| :x: | 78a463e7 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/10/) | 2022-12-08 21:49:07 | ~31 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/10/consoleText) |
| :x: | 78a463e7 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/10/) | 2022-12-08 21:49:15 | ~38 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/10/consoleText) |
| :x: | 78a463e7 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/10/) | 2022-12-08 21:49:32 | ~59 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/10/consoleText) |
| | | | | | | |
| :x: | 4f436fe0 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/11/) | 2022-12-09 15:46:18 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/11/consoleText) |
| :x: | 4f436fe0 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/11/) | 2022-12-09 15:46:24 | ~1 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/11/consoleText) |
| :x: | 4f436fe0 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/11/) | 2022-12-09 15:47:00 | ~2 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/11/consoleText) |
| | | | | | | |
| :heavy_check_mark: | bcb3da0e | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/12/) | 2022-12-09 16:23:54 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221209-162132-bcb3da-pr2962.zip) |
| :heavy_check_mark: | bcb3da0e | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/12/) | 2022-12-09 16:24:26 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221209-162136-bcb3da-pr2962.zip) |
| :heavy_check_mark: | bcb3da0e | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/12/) | 2022-12-09 16:25:42 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221209-162132-bcb3da-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | dd3cbd8f | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/13/) | 2022-12-12 21:24:45 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221212-211949-dd3cbd-pr2962.zip) |
| :heavy_check_mark: | dd3cbd8f | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/13/) | 2022-12-12 21:25:16 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221212-211945-dd3cbd-pr2962.zip) |
| :heavy_check_mark: | dd3cbd8f | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/13/) | 2022-12-12 21:26:03 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221212-211945-dd3cbd-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | a46d69fb | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/14/) | 2022-12-12 21:48:10 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221212-214255-a46d69-pr2962.zip) |
| :heavy_check_mark: | a46d69fb | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/14/) | 2022-12-12 21:49:42 | ~6 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221212-214255-a46d69-pr2962.zip) |
| :heavy_check_mark: | a46d69fb | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/14/) | 2022-12-12 21:50:18 | ~7 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221212-214255-a46d69-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | c8623bda | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/15/) | 2022-12-13 22:21:05 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221213-221902-c8623b-pr2962.zip) |
| :heavy_check_mark: | c8623bda | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/15/) | 2022-12-13 22:22:22 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221213-221858-c8623b-pr2962.aar) |
| :heavy_check_mark: | c8623bda | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/15/) | 2022-12-13 22:23:11 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221213-221858-c8623b-pr2962.zip) |
| | | | | | | |
| :heavy_check_mark: | 6690533c | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/16/) | 2022-12-14 16:48:00 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221214-164548-669053-pr2962.zip) |
| :heavy_check_mark: | 6690533c | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/16/) | 2022-12-14 16:48:19 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221214-164544-669053-pr2962.zip) |
| :heavy_check_mark: | 6690533c | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/16/) | 2022-12-14 16:49:23 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221214-164544-669053-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | 10233427 | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/17/) | 2022-12-15 16:47:27 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221215-164512-102334-pr2962.zip) |
| :heavy_check_mark: | 10233427 | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/17/) | 2022-12-15 16:47:35 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221215-164512-102334-pr2962.zip) |
| :heavy_check_mark: | 10233427 | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/17/) | 2022-12-15 16:50:33 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221215-164512-102334-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | da05530c | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/18/) | 2022-12-15 17:08:36 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221215-170635-da0553-pr2962.zip) |
| :heavy_check_mark: | da05530c | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/18/) | 2022-12-15 17:08:50 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221215-170635-da0553-pr2962.zip) |
| :heavy_check_mark: | da05530c | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/18/) | 2022-12-15 17:09:57 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221215-170631-da0553-pr2962.aar) |
| | | | | | | |
| :x: | d6144b11 | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/19/) | 2022-12-15 22:14:47 | ~36 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/19/consoleText) |
| :x: | d6144b11 | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/19/) | 2022-12-15 22:14:54 | ~43 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/19/consoleText) |
| :x: | d6144b11 | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/19/) | 2022-12-15 22:16:14 | ~2 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/19/consoleText) |
| | | | | | | |
| :heavy_check_mark: | 8f25ef84 | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/20/) | 2022-12-16 21:02:33 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221216-210031-8f25ef-pr2962.zip) |
| :heavy_check_mark: | 8f25ef84 | [#20](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/20/) | 2022-12-16 21:02:49 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221216-210026-8f25ef-pr2962.zip) |
| :heavy_check_mark: | 8f25ef84 | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/20/) | 2022-12-16 21:03:56 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221216-210026-8f25ef-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | 0736e134 | [#21](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/21/) | 2022-12-18 17:28:29 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221218-172551-0736e1-pr2962.zip) |
| :heavy_check_mark: | 0736e134 | [#21](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/21/) | 2022-12-18 17:30:11 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221218-172551-0736e1-pr2962.zip) |
| :heavy_check_mark: | 0736e134 | [#21](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/21/) | 2022-12-18 17:31:47 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221218-172547-0736e1-pr2962.aar) |
| | | | | | | |
| :heavy_check_mark: | a27360d7 | [#22](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/22/) | 2022-12-19 12:14:41 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221219-121144-a27360-pr2962.zip) |
| :heavy_check_mark: | a27360d7 | [#22](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/22/) | 2022-12-19 12:16:31 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221219-121144-a27360-pr2962.zip) |
| :heavy_check_mark: | a27360d7 | [#22](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/22/) | 2022-12-19 12:16:52 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221219-121144-a27360-pr2962.aar) |
| | | | | | | |
| :x: | 5aa8c82e | [#23](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/23/) | 2022-12-19 14:14:55 | ~27 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/23/consoleText) |
| :heavy_check_mark: | 5aa8c82e | [#23](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/23/) | 2022-12-19 14:16:19 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221219-141427-5aa8c8-pr2962.zip) |
| :heavy_check_mark: | 5aa8c82e | [#23](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/23/) | 2022-12-19 14:19:01 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221219-141427-5aa8c8-pr2962.aar) |
| | | | | | | |
| :x: | 33cda1d9 | [#24](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/24/) | 2022-12-19 21:12:15 | ~16 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/24/consoleText) |
| :heavy_check_mark: | 33cda1d9 | [#24](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/24/) | 2022-12-19 21:12:57 | ~58 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221219-211158-33cda1-pr2962.zip) |
| :heavy_check_mark: | 33cda1d9 | [#24](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/24/) | 2022-12-19 21:15:24 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221219-211158-33cda1-pr2962.aar) |
| | | | | | | |
| :x: | c036963d | [#25](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/25/) | 2022-12-20 14:40:23 | ~14 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/25/consoleText) |
| :heavy_check_mark: | c036963d | [#25](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/25/) | 2022-12-20 14:41:34 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221220-144007-c03696-pr2962.zip) |
| :heavy_check_mark: | c036963d | [#25](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/25/) | 2022-12-20 14:43:59 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221220-144007-c03696-pr2962.aar) |
| | | | | | | |
| :x: | 5de8ead7 | [#26](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/26/) | 2022-12-20 14:47:16 | ~25 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/26/consoleText) |
| :heavy_check_mark: | 5de8ead7 | [#26](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/26/) | 2022-12-20 14:48:11 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221220-144651-5de8ea-pr2962.zip) |
| :heavy_check_mark: | 5de8ead7 | [#26](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/26/) | 2022-12-20 14:50:45 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221220-144650-5de8ea-pr2962.aar) |
| :x: | 5de8ead7 | [#27](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/27/) | 2022-12-20 14:58:52 | ~11 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/27/consoleText) |
| | | | | | | |
| :x: | 2a9bad84 | [#28](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/28/) | 2022-12-29 12:11:38 | ~16 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/28/consoleText) |
| :heavy_check_mark: | 2a9bad84 | [#27](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/27/) | 2022-12-29 12:12:37 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221229-121120-2a9bad-pr2962.zip) |
| :heavy_check_mark: | 2a9bad84 | [#27](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/27/) | 2022-12-29 12:15:21 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221229-121116-2a9bad-pr2962.aar) |
| | | | | | | |
| :x: | ef3d2239 | [#29](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/29/) | 2022-12-29 13:00:09 | ~15 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/29/consoleText) |
| :heavy_check_mark: | ef3d2239 | [#28](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/28/) | 2022-12-29 13:01:23 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221229-125956-ef3d22-pr2962.zip) |
| :heavy_check_mark: | ef3d2239 | [#28](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/28/) | 2022-12-29 13:03:14 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221229-125952-ef3d22-pr2962.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 66c135aa | [#29](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/29/) | 2023-01-04 09:54:08 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230104-095251-66c135-pr2962.zip) |
| :heavy_multiplication_x: | 66c135aa | [#31](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/31/) | 2023-01-04 09:55:03 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/31//consoleText) |
| :heavy_check_mark: | 66c135aa | [#29](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/29/) | 2023-01-04 09:56:08 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230104-095248-66c135-pr2962.aar) |
| :heavy_check_mark: | 66c135aa | [#30](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/30/) | 2023-01-04 10:04:04 | ~11 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230104-095248-66c135-pr2962.zip) |
| | | | | | | |
| :heavy_check_mark: | 63efa6ae | [#30](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2962/30/) | 2023-01-04 10:26:10 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230104-102159-63efa6-pr2962.zip) |
| :heavy_check_mark: | 63efa6ae | [#30](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2962/30/) | 2023-01-04 10:26:53 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230104-102159-63efa6-pr2962.aar) |
| :heavy_check_mark: | 63efa6ae | [#31](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2962/31/) | 2023-01-04 10:26:58 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230104-102159-63efa6-pr2962.zip) |
| :heavy_multiplication_x: | 63efa6ae | [#32](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/32/) | 2023-01-04 10:39:32 | ~17 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/32//consoleText) |
| :heavy_multiplication_x: | 63efa6ae | [#33](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/33/) | 2023-01-04 10:46:43 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/33//consoleText) |
| :heavy_check_mark: | 63efa6ae | [#34](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/34/) | 2023-01-05 07:38:12 | ~8 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-2962/34//consoleText) |

alaibe commented on appdatabase/migrations/sql/1671451982_add_balance_history.up.sql:
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE IF NOT EXISTS balance_history (
+    chain_id UNSIGNED BIGINT NOT NULL,
+    address VARCHAR NOT NULL,
+    currency VARCHAR NOT NULL,
```
I believe we should always store the value in USD, and convert only if needed.
WDYT?

stefandunca replied:
I use this to identify the currency for which the balance was cached (ISO), which, along with chain id and address, is part of the unique balance identity. This is the way to get the ground truth, and any other value reinterpretation will add errors and complexity that may propagate and generate side effects if we want to do anything other transformations.

Your example will work only if we want to show also the history of the currency converted to USD. That means showing the USD value at the time and not accounting for inflation. We want to display the historical amount with the current conversion (adjusted for inflation).

Thanks for the heads up; I check this point of view with the design.

alaibe replied:
Good point, regarding the difference between various currency. The trade off is that it will cost more request, I would imagine that showing the value according to current currency values is what makes more sense.

But yes better check with John

alaibe commented on appdatabase/migrations/sql/1671451982_add_balance_history.up.sql:
```diff
@@ -0,0 +1,15 @@
+    timestamp INT NOT NULL,
+    bitset INT NOT NULL,
+    balance BLOB
+);
```
Shall we also add token_address ?

stefandunca replied:
Good question, thanks for the heads up. My understanding is that the short currency is standardized and enough. The `token_address` for the contract is retrieved from our token store. Now I wonder if this can change. Can it be that the contract address for an ERC20 token can change without changing the short currency? Will ask around and look into.

alaibe commented on appdatabase/migrations/sql/1671451982_add_balance_history.up.sql:
```diff
@@ -0,0 +1,15 @@
+    block BIGINT NOT NULL,
+    timestamp INT NOT NULL,
+    bitset INT NOT NULL,
+    balance BLOB
```
Why a blob? Can we store the hex value?

stefandunca replied:
This is more efficient and simple, and I have seen it done in other parts of the code for `big.Int`s. I can't see how this will improve other than human readability, which I didn't consider.

alaibe replied:
The only cons of this is that it is not possible to query against this value correct?

stefandunca replied:
The query should work with byte comparison (I can't see any case where it is required for balance, and we don't index it, though)

alaibe commented on protocol/protobuf/chat_message.pb.go:
```diff
@@ -161,7 +161,6 @@ func (m *StickerMessage) GetPack() int32 {
 type ImageMessage struct {
 	Payload              []byte    `protobuf:"bytes,1,opt,name=payload,proto3" json:"payload,omitempty"`
 	Type                 ImageType `protobuf:"varint,2,opt,name=type,proto3,enum=protobuf.ImageType" json:"type,omitempty"`
-	AlbumID   			 string    `protobuf:"bytes,3,opt,name=album_id,json=albumId,proto3" json:"album_id,omitempty"`
```
I am surprise to see modification here?

stefandunca replied:
I'm also surprised to see it; I didn't check what was generated by the `make generate`. Will look closer and clarify this.

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+}
+
+func (bh *BalanceHistory) StartBalanceHistory(rpcClient *rpc.Client, networkManager *network.Manager, tokenManager *token.Manager) {
+	go func() {
```
We need to be able to stop it when the wallet stop. you can look at how to do it in the StartWallet from the API.
Also shall this is not be triggered when we start the wallet?

stefandunca replied:
The task is cancelable and is done by the `PriorityBasedAsyncQueue` when a request for an intermediate balance report comes in, not to run the short update with the complete update simultaneously. However, I don't stop it if the wallet service stops.

I decided this after inquiring if we ever stop the wallet after the service is started and learning that we don't.

Now I'm wondering what happens when we try to close the app tough. I will look into it to be sure we gracefully stop the current update process.

alaibe replied:
Yes this is better to ensure that the wallet can be stopped, this is how we it in other part, also we need to consider mobile which I believe stop the wallet

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+)
+
+var secondsInTimeInterval = map[BalanceHistoryTimeInterval]int64{
+	BalanceHistory7Hours:  7 * 24 * 60 * 60,
```
Why not using Duration?

stefandunca replied:
I wasn't considering this at that time. However, all the latest improvements are using duration. I will also improve on this with the latest improvements PR (WIP) if the type transition doesn't incur any unnecessary conversions.

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+	BalanceHistoryAllTime: 50 * 365 * 24 * 60 * 60,
+}
+
+const secondsInHour = 60 * 60
```
Why not using duration?

stefandunca replied:
I will fix this in the current improvements PR to have fewer changes to rebase on.

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+
+func (bh *BalanceHistory) getDBBalanceEntriesByTimeIntervalAndSortedAsc(identify *bhIdentity, startingAtBlock *big.Int, filter *bhFilter, maxEntries int) ([]*balanceHistoryDBEntry, error) {
+	return bh.getDBBalanceEntriesTimeSorted(identify, startingAtBlock, filter, maxEntries, true)
+}
```
More like a personal opinion: those functions are useless, they bring very little to no value, but just confuse me.

If parameters are complex we can create a struct to pass to the function

If it is too much, in this context, you could create 2 type, for the orders: Asc and Desc

stefandunca replied:
Actually, readability was the reason why I added them. Good to know it didn't work out, and a type alias is a better approach. I will prefer to fix it in the follow-up improvements PR not to rebase all these changes on the ERC20 PR also.

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+		queryStr = "SELECT block, timestamp, balance FROM balance_history WHERE chain_id = ? AND address = ? AND currency = ? AND block >= ? AND timestamp BETWEEN ? AND ? AND (bitset & ?) > 0 ORDER BY block ASC LIMIT ?"
+	} else {
+		queryStr = "SELECT block, timestamp, balance FROM balance_history WHERE chain_id = ? AND address = ? AND currency = ? AND block >= ? AND timestamp BETWEEN ? AND ? AND (bitset & ?) > 0 ORDER BY block DESC LIMIT ?"
+	}
```
no need to rewrite the whole string, using `fmt.Sprint` function should do it

stefandunca replied:
This is how it was done initially, but I got warnings from our security check-in CI that using formatting introduces security risks and should be avoided.

stefandunca replied:
Not sure how I can do a secure string formatting and be sure the formated values are correctly escaped.

alaibe replied:
we use `fmt.Sprintf ` in many other places, i think we should not duplicate that string

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+type bhIdentity struct {
+	chainID  uint64
+	address  common.Address
+	currency string
```
I think it is a mistake to store the currency, this will be simple multiplication to convert to the right one, storing to a default one seems much simpler

stefandunca replied:
If I do this, I also have to store the conversion rate to have the value in another currency or shown at the current conversion rate, which looks unnecessary. But I will have a better view of this matter after discussing with the design if we need to show the balance history with the current conversion rate or keep the historical one.

alaibe replied:
agree it all depends of: `we need to show the balance history with the current conversion rate or keep the historical one`

alaibe commented on services/wallet/transfer/balance_history.go:
```diff
@@ -0,0 +1,728 @@
+package transfer
```
This whole balance history deserves its own package to me, it is quite complex feature. Then we could split into multiple files for readbility: e.g: db.go

stefandunca replied:
I agre but will like to have this done in a separate follow up PR.

alaibe replied:
this makes it super hard to review, please consider this

alaibe reviewed (commented):
Maybe I am totally missing something but the logic to ensure caching seems very complex. Wouldn't it be simpler to:
1) Assume block length is always 12s.
2) Store in cache the block number and the time
3) if we need to go back in time 2 days, because last cache was 2 days ago, we calculate 2 point in time for the 2 days. and we fetch the data by adding constant block number

to continue on the example:
last record in cache is:
block number is 100
date is 6th jan at 8:00 UTC

block interval  is 86400 / 15 => so fetch 5760 + 100 and 5760 *2+100

stefandunca commented:
> Maybe I am totally missing something but the logic to ensure caching seems very complex. Wouldn't it be simpler to:

I agree with you and also thought of why something I estimated to be so simple ended up so complicated. It is a combination of the following factors:
- We don't have a centralized environment to optimize filtering and to cache of balance change events as they happen.
- The requirements with which I started were assuming that I could rely on the transaction history, then evolved that the user can fetch at any time with varying time intervals for any account that can have internal transactions done outside our system.
- My shallow experience with blockchain development and taking the requirements (like 12s fix block interval target literally)

With the newly learned lessons and new long-time caching and flexible sampling requirements, I can see how to simplify the logic, but that will mean a complete rewrite of the fetch and caching, which already works. I would like to see how it goes in the field and prioritize the simplified logic rewrite if we have issues with the current logic. Let's discuss this in the next meeting.

> 1. Assume block length is always 12s.

That is how I started based on the information that the target block time is 12 seconds. However, I had substantial errors that accumulated to generate offsets of days and weeks of missing data for some intervals.
The errors seemed too much to me. After more investigations, I discovered that the block time ranges from 12s to 15s and varies significantly from block to block depending on demand and other factors.

Therefore, I added the calibration part that fetches a few more block timestamps for each blockchain to compensate for errors that significantly improved the results. This complicated the code, but I added tests to support it.

> 2. Store in cache the block number and the time

I store this info along with the identity part (blockchain, currency and address) and a bitset marker to keep the entry associated with the period for quicker fetching.

> 3. if we need to go back in time 2 days, because last cache was 2 days ago, we calculate 2 point in time for the 2 days. and we fetch the data by adding constant block number
> to continue on the example: last record in cache is: block number is 100 date is 6th jan at 8:00 UTC
> 
> block interval is 86400 / 15 => so fetch 5760 + 100 and 5760 *2+100

You described how the implementation was initially done, but I was still missing cache hits due to errors introduced by variations in the block length combined with going back in time. Hence to stabilize the estimations, I had to start fetching from the first block and go up to have a reproducible pattern.
These changes were still insufficient because the start block was varying; hence the errors in the subsequent blocks were propagating to accumulate and have offsets between the blocks. I improved on this by adding the time snapping.

alaibe commented:
![image](https://user-images.githubusercontent.com/491074/211320030-015b296d-1362-4432-975d-daba0c45ff63.png)

FYI this is why I am saying we can consider block are produced every 12s, this is very stable now.

I still don't understand how you will miss cache hit if you start from latest cache?

stefandunca commented:
> ...
> FYI this is why I am saying we can consider block are produced every 12s, this is very stable now.

From which chain is the graph? It is the stable part after transitioning to PoS? Ca we rely on the fact that all blockchains we support will have this stability? In my tests, I haven't seen a stable one.

> I still don't understand how you will miss cache hit if you start from latest cache?

When computing the next block with a fixed block length, we deviate (miss the cache) with one block as soon as the error sum of all the blocks you step over accumulates close to one block length (e.g. 12s and depends on the rounding factor)

For example, in the 1-year time interval, with 12s block length at 100 samples, there will be 26298 blocks in one stride. That means the error of each block should be under ~0.00045 seconds or to vary up and down in a stable manner to compensate for this, which is improbable.

stefandunca commented:
I will simplify the algorithm as we discussed. Given the recent change in user requirements, the precision and flexibility offered by the current implementation might not be worth the maintenance burden.

stefandunca commented:
Closing in favour of the simplification refactoring: https://github.com/status-im/status-go/pull/3071
